include/alp/algorithms/householder_qr.hpp

```diff
@@ -0,0 +1,244 @@
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+
+#include <alp.hpp>
+#include <graphblas/utils/iscomplex.hpp> // use from grb
+#include "../tests/utils/print_alp_containers.hpp"
+// TEMPDISABLE enables workarounds for non-implemented features
+// should be removed in the final version
+#define TEMPDISABLE
+
+namespace alp {
+
+	namespace algorithms {
+
+		/**
+		 * @brief Computes Householder QR decomposition of general matrix \f$H = QR\f$
+		 *        where \a H is general (complex or real),
+		 *        \a R is upper triangular (if H is not square,
+		 *        R is of the same shape with zeros below diagonal), and
+		 *        \a Q is orthogonal.
+		 *
+		 * @tparam D        Data element type
+		 * @tparam Ring     Type of the semiring used in the computation
+		 * @tparam Minus    Type minus operator used in the computation
+		 * @tparam Divide   Type of divide operator used in the computation
+		 * @param[out] Q    output orthogonal matrix such that H = Q T Q^T
+		 * @param[out] R    output same shape as H with zeros below diagonal
+		 * @param[in]  H    input general matrix
+		 * @param[in]  ring A semiring for operations
+		 * @return RC       SUCCESS if the execution was correct
+		 *
+		 */
+		template<
+			typename D,
+			typename GeneralType,
+			typename OrthogonalType,
+			class Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			class Minus = operators::subtract< D >,
+			class Divide = operators::divide< D >
+		>
+		RC householder_qr(
+			Matrix< D, GeneralType, Dense > &H,
+			Matrix< D, OrthogonalType, Dense > &Q,
+			Matrix< D, GeneralType, Dense > &R,
```

Comment:
If this algorithm is called on matrix views, the compiler would throw no matching call error because the default view type does not match the value used in matrix view. Therefore, View, ImfR and ImfC should be template parameters.
Additionally, since the algorithms can be compiled for different backends, the backend should also be a template parameter.

Reply:
I'm not sure if structure::Orthogonal is the most general structure we can have. In principle we can non-full-rank orthogonal matrix here, something we need to discuss. I'll skip the static_assert at the moment if that is OK @hidanielesgit  ?